ethcore/src/state/mod.rs

```diff
@@ -2196,6 +2266,27 @@ mod tests {
 		assert_eq!(state.balance(&a).unwrap(), U256::from(0));
 	}
 
+	#[test]
+	fn checkpoint_get_storage_at() {
+		let mut state = get_temp_state();
+		let a = Address::zero();
+		let k = H256::from(U256::from(0));
+		let c1 = state.checkpoint();
+		state.set_storage(&a, k, H256::from(U256::from(1))).unwrap();
+		let c2 = state.checkpoint();
+		let c3 = state.checkpoint();
+		state.set_storage(&a, k, H256::from(U256::from(3))).unwrap();
+		let c4 = state.checkpoint();
+		state.set_storage(&a, k, H256::from(U256::from(4))).unwrap();
+		let c5 = state.checkpoint();
```

Comment:
Can you check what happens when checkpoints are discarded / committed? I believe checkpoints represent:
"The changes that has to be made to the cache to return to a point in time." and because of that the logic at `checkpoint_storage_at` might be off.

Reply:
Here's some added tests for discard/revert of checkpoints. https://github.com/paritytech/parity-ethereum/pull/9319/commits/c2b9e9741cb9d92c48eab2c200ba33f11de0290a
As @tomusdrw pointed out, we probably still need to check coverage to make sure all branches are tested.